Parser/pegen.c

```diff
@@ -2383,28 +2383,90 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
         return _PyAST_Constant(res, NULL, lineno, col_offset, end_lineno, end_col_offset, p->arena);
     }
 
-    asdl_expr_seq* values = _Py_asdl_expr_seq_new(n_elements, p->arena);
-    if (values == NULL) {
+    if (!f_string_found && len == 1) {
+        return asdl_seq_GET(strings, 0);
+    }
+
+    asdl_expr_seq* flattened = _Py_asdl_expr_seq_new(n_flattened_elements, p->arena);
+    if (flattened == NULL) {
         return NULL;
     }
 
+    /* build flattened list */
     Py_ssize_t current_pos = 0;
     Py_ssize_t j = 0;
     for (i = 0; i < len; i++) {
         expr_ty elem = asdl_seq_GET(strings, i);
         if (elem->kind == Constant_kind) {
-            asdl_seq_SET(values, current_pos++, elem);
+            asdl_seq_SET(flattened, current_pos++, elem);
         } else {
             for (j=0; j < asdl_seq_LEN(elem->v.JoinedStr.values); j++) {
                 expr_ty subvalue = asdl_seq_GET(elem->v.JoinedStr.values, j);
                 if (subvalue == NULL) {
                     return NULL;
                 }
-                asdl_seq_SET(values, current_pos++, subvalue);
+                asdl_seq_SET(flattened, current_pos++, subvalue);
+            }
+        }
+    }
+
+    /* calculate folded element count */
+    Py_ssize_t n_elements = 0;
+    int prev_is_constant = 0;
+    for (i = 0; i < n_flattened_elements; i++) {
+        expr_ty elem = asdl_seq_GET(flattened, i);
+        if (!(prev_is_constant && elem->kind == Constant_kind)) {
+            n_elements++;
+        }
+        prev_is_constant = elem->kind == Constant_kind;
+    }
+
+    asdl_expr_seq* values = _Py_asdl_expr_seq_new(n_elements, p->arena);
+    if (values == NULL) {
+        return NULL;
+    }
+
+    /* build folded list */
+    _PyUnicodeWriter writer;
+    current_pos = 0;
+    for (i = 0; i < n_flattened_elements; i++) {
+        expr_ty elem = asdl_seq_GET(flattened, i);
+        if (elem->kind == Constant_kind) {
+            _PyUnicodeWriter_Init(&writer);
+            expr_ty last_elem = elem;
+            for (j = i; j < n_flattened_elements; j++) {
+                elem = asdl_seq_GET(flattened, j);
+                if (elem->kind == Constant_kind) {
+                    if (_PyUnicodeWriter_WriteStr(&writer, elem->v.Constant.value)) {
+                        return NULL;
+                    }
+                    last_elem = elem;
+                } else {
+                    break;
+                }
+            }
```

Comment:
There is an optimization to be done here, we could skip `_PyUnicodeWriter` and creating a new AST constant altogether if it's just one string. I did not do it initially for simplicity's sake, should I do it now?

Reply:
Yeah, it makes sense because that's like the dominant path for f-strings.

Reply:
But be carefull, we are already burning a lot of indentation levels and this code is already difficult to read

Reply:
Ideally you can do an early return close to the exit for 1 string that is not an f-string before:

```
    if (!f_string_found && len == 1) {
        return asdl_seq_GET(strings, 0);
    }
```

Reply:
I already do that optimization, see the 2nd commit.

> But be careful, we are already burning a lot of indentation levels and this code is already difficult to read

With my refactor, this is now actually much easier to do, I can just add some clausules to the if that decides if we for for the `_PyUnicodeWriter` branch.

Reply:
I added a comment and some spacing to make the code easier to digest.

Reply:
> I already do that optimization, see the 2nd commit.

I know you do, I am suggesting to do the one you are describing now close to tha tone

Reply:
Ah, yes, sorry. I just needed to add `i+1 < n_flattened_elements && asdl_seq_GET(flattened, i+1)->kind == Constant_kind` to the `if`.